### DISCUSSION TITLE
Get DISTINCT Statements

### DISCUSSION THREAD

**Discussion by naturzukunft** [eclipse-rdf4j/rdf4j#4518] [Q&A] [answered]
Hi there,
I am once again at a loss.
i get a model out of rdf4j 
`Model resultModel = QueryResults.asModel(con.getStatements(subject, null, null, contexts.toArray(new Resource[] {})));`
some statements are now duplicated in the model because the subject occurs in multiple contexts.

Ho can i get the DISTINCT (by subject & predicate) statements?

Thanks Fredy

**Comment by abrokenjester** [comment, marked as answer]
The main thing to understand is that if you _ask_ for statements from more than one context, then you _get_ statements from more than one context. In RDF4J, the context is part of the identity of a statement, so the same (subject, predicate, object) in multiple contexts are _not_ considered identical to each other.

Assuming your intent is to get a Model that doesn't _have_ contexts (and therefore no 'duplicated' triples), then one way to fix is by using a SPARQL query instead of a simple getStatement operation, to control the exact shape of the returned data. Something like this:

```sparql
CONSTRUCT { ?s ?p ?o } 
FROM <context1>
FROM <context2>
...
WHERE { ?s ?p ?o }
```

Another way is to convert statements on the fly, as you're retrieving, before sticking them into your model. Something like this:

```java
Model resultModel = DynamicModelFactory.createEmptyModel();
con.getStatements(subject, null, null, contexts.toArray(new Resource[] {}))
    .forEach(st -> resultModel.add(st.getSubject(), st.getPredicate(), st.getObject()));
```

You can also do a similar trick with your original `resultModel` of course, if you want to keep the original "with context" result around, e.g. something like this:

```java
Model resultModel = QueryResults.asModel(con.getStatements(subject, null, null, contexts.toArray(new Resource[] {})));

Model distinctTriples = resultModel.stream()
    .map(st -> statement(st.getSubject(), st.getPredicate(), st.getObject(), null))
    .collect(ModelCollector.toModel());
```

**Reply by hmottestad** [reply]
There is also the static method `Statements.stripContext(st)` that functions the same way under the hood, but helps to add some clarity to the intentions of the code.

```
Model distinctTriples = resultModel.stream()
    .map(Statements::stripContext)
    .collect(ModelCollector.toModel());
```

https://rdf4j.org/javadoc/latest/org/eclipse/rdf4j/model/util/Statements.html#stripContext(org.eclipse.rdf4j.model.Statement)

**Reply by hmottestad** [reply]
I would also recommend always using try-with-resource, to be on the safe side.

```
try(var iteration = con.getStatements(subject, null, null, contexts.toArray(new Resource[] {}))){
    Model resultModel = QueryResults.asModel(iteration);
}
```

**Reply by abrokenjester** [reply]
> I would also recommend always using try-with-resource, to be on the safe side.
> 
> ```
> try(var iteration = con.getStatements(subject, null, null, contexts.toArray(new Resource[] {}))){
>     Model resultModel = QueryResults.asModel(iteration);
> }
> ```

That's not actually necessary. `QueryResults#asModel` guarantees that the input iteration will be closed, regardless of whether it is fully consumed or an exception occurred.

**Reply by hmottestad** [reply]
> > I would also recommend always using try-with-resource, to be on the safe side.
> 
> > 
> 
> > ```
> 
> > try(var iteration = con.getStatements(subject, null, null, contexts.toArray(new Resource[] {}))){
> 
> >     Model resultModel = QueryResults.asModel(iteration);
> 
> > }
> 
> > ```
> 
> 
> 
> That's not actually necessary. `QueryResults#asModel` guarantees that the input iteration will be closed, regardless of whether it is fully consumed or an exception occurred. 

The code delegates closing to the addAll(...) method. There are several method calls before then. If one of those should fail then the iteration would not be closed. I don't think any of those methods will throw exceptions, so it should be safe to assume that the iteration will be closed. 

I would still recommend to not assume that some other code will close your resources for you. I've been caught out with stuff like null checks in constructors that result in exceptions where I wouldn't have expected any. (e.g. new A(new B(new C(...))).close() would leave C and B unclosed if A fails in the constructor). We also have some services based code in RDF4J where you would get a ClassNotFoundException because of a missing maven dependency, but there would be no compile time errors like you would typically get when a dependency is missing.
